implementations/rust/ockam/ockam_api/src/nodes/service/forwarder.rs

```diff
@@ -49,14 +94,135 @@ impl NodeManager {
+    let mut protos = input.iter();
+    if !matches!(protos.next().map(|p| p.code()), Some(DnsAddr::CODE | Ip4::CODE | Ip6::CODE)) {
+        return None;
+    }
+    if !matches!(protos.next().map(|p| p.code()), Some(Tcp::CODE)) {
+        return None;
+    }
+    if let Some(p) = protos.next() {
+        if let Some(p) = p.cast::<Service>() {
+            if &*p == "api" {
+                return MultiAddr::default().try_with(input.iter().take(3)).ok();
+            }
+        }
+    }
+    None
+}
+
+/// Create a secure channel to an API node or return the address route as is.
+async fn connect(
+    manager: &mut NodeManager,
+    addr: &MultiAddr,
+    auth: &HashMap<MultiAddr, IdentityIdentifier>,
+    mode: CredentialExchangeMode,
+) -> Result<Either<Address, Route>> {
+    if let Some(a) = api_host(addr) {
+        if let Some(i) = auth.get(&a) {
+            let r = multiaddr_to_route(&a).ok_or_else(|| ApiError::generic("invalid multiaddr"))?;
+            let i = Some(vec![i.clone()]);
+            let a = manager.create_secure_channel_impl(r, i, mode, None).await?;
+            return Ok(Either::Left(a));
+        } else {
+            info!(%addr, "no authorised identity found")
+        }
+    }
+    let r = multiaddr_to_route(addr).ok_or_else(|| ApiError::generic("invalid multiaddr"))?;
+    Ok(Either::Right(r))
+}
+
+/// Configure the session for automatic recovery.
+fn enable_recovery(
+    session: &mut Session,
+    manager: Address,
+    ctx: Arc<Context>,
+    addr: MultiAddr,
+    alias: String,
+    auth: Arc<HashMap<MultiAddr, IdentityIdentifier>>,
+    mode: CredentialExchangeMode,
+) {
+    session.set_replacement(move |prev| {
+        let ctx = ctx.clone();
+        let addr = addr.clone();
+        let alias = alias.clone();
+        let auth = auth.clone();
+        let manager = manager.clone();
+        Box::pin(async move {
+            debug!(%addr, "creating new remote forwarder");
+            let f = async {
+                let a = replace_sec_chan(&ctx, &manager, prev, &addr, &auth, mode).await?;
+                RemoteForwarder::create_static(&ctx, Route::from(a.clone()), alias).await?;
+                Ok(a)
+            };
+            match timeout(MAX_RECOVERY_TIME, f).await {
+                Err(_) => {
+                    warn!(%addr, "timeout creating new remote forwarder");
+                    Err(ApiError::generic("timeout"))
+                }
+                Ok(Err(e)) => {
+                    warn!(%addr, err = %e, "error creating new remote forwarder");
+                    Err(e)
+                }
+                Ok(Ok(a)) => Ok(a),
+            }
+        })
+    })
+}
+
+async fn replace_sec_chan(
+    ctx: &Context,
+    manager: &Address,
+    prev: Address,
+    addr: &MultiAddr,
+    authorised: &HashMap<MultiAddr, IdentityIdentifier>,
+    mode: CredentialExchangeMode,
+) -> Result<Address> {
+    debug!(%addr, %prev, "recreating secure channel");
```

Comment:
Is sending a message to itself is better that simple function call?

Reply:
Most methods of `NodeManager` require unique access (`&mut self`). The recovery futures run asynchronously and we would have to wrap the manager into an `Arc<Mutex<_>>` to apply those methods. Such a change would be much larger so I chose the message sending instead.

Reply:
Yeah, probably there should be `NodeManager` that is doing all the heavy lifting, and `NodeManagerWorker` that is a Worker that delegates calls to `NodeManager`. In this case either `NodeManager` can be wrapped inside `Arc<Mutex>` of its fields could be turned into thread-safe structures. Definitely out of the context of this PR though. Thoughts?